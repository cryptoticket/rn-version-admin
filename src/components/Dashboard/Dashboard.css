.app__global-loader {
	position: absolute;
    top: 70px;
    right: 20px;
}

.content {
	margin: 0;
}
