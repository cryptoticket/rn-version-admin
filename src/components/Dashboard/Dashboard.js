import { Button } from 'primereact/button';
import { Toolbar } from 'primereact/toolbar';
import React, { setGlobal } from 'reactn';
import { BrowserRouter as Router, Switch, Route, Link, Redirect } from 'react-router-dom';
import { Login, Version } from '../../pages';
import './Dashboard.css';

// set an initial global state
setGlobal({
	redirect: null,
	user: null
});

/**
 * Component renders a dashboard with top menu and navigation
 */
export default class Dashboard extends React.Component {
	/**
	 * On component init
	 */
	componentDidMount() {
		// if oauth token was passed then save it to local storage
		const queryParams = new URLSearchParams(window.location.search);
		const queryToken = queryParams.get('token');
		if(queryToken) localStorage.setItem('token', queryToken);
		// if token exists in local storage
		const localToken = localStorage.getItem('token');
		if(localToken) {
			// pass token to global state and redirect user to versions page
			this.setGlobal({ 
				redirect: '/versions',
				user: { token: localToken }
			});
		} else {
			// user is not authorized, redirect him to login page
			this.setGlobal({ 
				redirect: '/'
			});
		}
	};

	/**
	 * On "logout" button click
	 */
	onLogoutClick = () => {
		// delete token from local storage
		localStorage.removeItem('token');
		// clear user from global state and redirect user to login page
		this.setGlobal({ 
			redirect: '/',
			user: null
		});
	};

	/**
	 * Renders JSX template
	 * @return {Object} JSX template
	 */
	render() {
		return (
			<Router>
				<div className="p-5">
					{/* toolbar */}
					<div>
						<Toolbar>
							<div className="p-toolbar-group-left">
								<strong>React Native Version Admin</strong>
							</div>
							{/* show top menu only to authorized user */}
							{this.global.user && 
								<div className="p-toolbar-group-right">
									<Link to="/versions">
										<Button label="Versions" icon="pi pi-cloud" className="p-button-secondary" />
									</Link>
									<Button label="Logout" icon="pi pi-power-off" className="p-button-secondary" onClick={this.onLogoutClick} />
								</div>
							}
						</Toolbar>
					</div>
					{/* main content */}
					<div className="p-grid content">
						<div className="p-col-12">
						<Switch>
							<Route path="/versions">
								<Version />
							</Route>
							<Route path="/">
								<Login />
							</Route>
						</Switch>
						</div>
					</div>
				</div>
				{/* app global redirect */}
				{ this.global.redirect && <Redirect to={this.global.redirect} /> }
			</Router>
		);
	}
}
