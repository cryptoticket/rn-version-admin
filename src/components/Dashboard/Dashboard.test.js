import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';

import Dashboard from './Dashboard';

configure({ adapter: new Adapter() });

describe('Dashboard component', () => {
    describe('componentDidMount()', () => {
		it('should set auth token if it exists', function () {
			delete window.location;
			window.location = {
				search: '?token=ANY_TOKEN'
			};
			const spyLocalStorageSetItem = jest.spyOn(Storage.prototype, 'setItem');
			const wrapper = shallow(<Dashboard />);
			const instance = wrapper.instance();
			instance.setGlobal = jest.fn();
			instance.componentDidMount();
			expect(spyLocalStorageSetItem).toHaveBeenCalled();
			expect(instance.setGlobal).toHaveBeenCalledWith({
				redirect: '/versions',
				user: { token: 'ANY_TOKEN' }
			});
			// remove created token from local storage
			localStorage.removeItem('token');
		});
		
		it('should redirect to login page if token does not exist', function () {
			window.location = {
				search: ''
			};
			const wrapper = shallow(<Dashboard />);
			const instance = wrapper.instance();
			instance.setGlobal = jest.fn();
			instance.componentDidMount();
			expect(instance.setGlobal).toHaveBeenCalledWith({
				redirect: '/'
			});
        });
	});

	describe('onLogoutClick()', () => {
		it('should delete auth token and redirect user to login page', function () {
			const spyLocalStorageRemoveItem = jest.spyOn(Storage.prototype, 'removeItem');
			const wrapper = shallow(<Dashboard />);
			const instance = wrapper.instance();
			instance.setGlobal = jest.fn();
			instance.onLogoutClick();
			expect(spyLocalStorageRemoveItem).toHaveBeenCalled();
			expect(instance.setGlobal).toHaveBeenCalledWith({
				redirect: '/',
				user: null
			});
		});
	});

	describe('render()', () => {
		it('should render component', function () {
			const component = renderer.create(<Dashboard global={{user: {id: 1}, redirect: '/any', isLoading: true}} />);
			expect(component.toJSON()).toMatchSnapshot();
		});
	});
});
