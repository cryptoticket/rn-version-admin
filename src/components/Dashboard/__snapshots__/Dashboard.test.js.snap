// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component render() should render component 1`] = `
Array [
  <div
    className="p-5"
  >
    <div>
      <div
        className="p-toolbar p-component"
        id={null}
        role="toolbar"
        style={null}
      >
        <div
          className="p-toolbar-group-left"
        >
          <strong>
            React Native Version Admin
          </strong>
        </div>
        <div
          className="p-toolbar-group-right"
        >
          <a
            href="/users"
            onClick={[Function]}
          >
            <button
              className="p-button p-component p-button-secondary p-button-text-icon-left"
            >
              <span
                className="pi pi-users p-c p-button-icon-left"
              />
              <span
                className="p-button-text p-c"
              >
                Users
              </span>
            </button>
          </a>
          <a
            href="/versions"
            onClick={[Function]}
          >
            <button
              className="p-button p-component p-button-secondary p-button-text-icon-left"
            >
              <span
                className="pi pi-cloud p-c p-button-icon-left"
              />
              <span
                className="p-button-text p-c"
              >
                Versions
              </span>
            </button>
          </a>
          <button
            className="p-button p-component p-button-secondary p-button-text-icon-left"
            onClick={[Function]}
          >
            <span
              className="pi pi-power-off p-c p-button-icon-left"
            />
            <span
              className="p-button-text p-c"
            >
              Logout
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      className="p-grid content"
    >
      <div
        className="p-col-12"
      >
        <div
          className="login"
        >
          <div
            className="p-card p-component login__card"
            style={null}
          >
            <div
              className="p-card-body"
            >
              <div
                className="p-card-content"
              >
                 
                <div
                  className="p-inputgroup p-justify-center"
                >
                  <button
                    className="p-button p-component p-button-text-icon-left"
                    onClick={[Function]}
                  >
                    <span
                      className="pi pi-google p-c p-button-icon-left"
                    />
                    <span
                      className="p-button-text p-c"
                    >
                      Login with Google
                    </span>
                  </button>
                </div>
                 
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="p-growl p-component p-growl-topright"
    id={null}
    style={null}
  >
    <div />
  </div>,
]
`;
